SOCAM integration

fedorgrab opened the pull request.

magrenimish committed:
socam integration trial 1

magrenimish committed:
Include SOCAM model and resources from SOCAM branch

sjfleming commented:
Let's add a test for socam and then also just make sure all the CI tests pass

`make format`
`make typecheck`
will help

sjfleming committed:
lint and mypy and add socam tests

sjfleming commented:
Todo: @magrenimish take a look at the unnormalized categorical dist.  Is the code implementation really unnormalized as intended?

sjfleming committed:
add tests for the unnormalized categorical: is there an issue?

sjfleming commented:
For now, let's leave the categorical implementation as it is.

BUT... here's my question @magrenimish : if we leave the categorical like it is written now (normalized), do we need to implement it ourselves?  Or can we just use Pyro's categorical?

EDIT [later]: it is fine the way it is written in the codebase currently. It does serve a function.

sjfleming committed:
add encoded targets transform

sjfleming commented:
Looks like we forgot the `Transform` called `EncodedTargets`.  I've added it back in.

fedorgrab committed:
Refactor SOCAM tests to match new model signature

- Remove pickle file mocking and GCS read_pkl_from_gcs patches
- Pass tensors and category lists directly to SOCAM constructor
- Add tests for compute_valid_mask function
- Use function-based pytest style instead of class-based

sjfleming committed:
Merge branch 'main' into socam_integration

sjfleming committed:
Merge branch 'main' into socam_integration

sjfleming commented:
Add socam test to test_cli

sjfleming committed:
add data utilities for cell ontology

sjfleming committed:
linting

sjfleming committed:
clean up code and inputs

sjfleming committed:
fix tests

sjfleming committed:
add socam cli test

sjfleming committed:
further tweaking

sjfleming committed:
compile prob_prop for inference

sjfleming committed:
remove pkl reader

sjfleming committed:
remove EncodedTargets for now

sjfleming committed:
add comment

sjfleming committed:
increase error tolerance to get to pass

sjfleming committed:
handle different subsets of cl_names for train and test

sjfleming committed:
refactor y_n to be cl_names_n which is string labels rather than integers

sjfleming committed:
change cl_name subsetting to the object level

magrenimish commented on cellarium/ml/models/socam.py:
```diff
@@ -0,0 +1,333 @@
+)
+
+
+def compute_valid_mask(input_categories: list[str], output_categories: list[str]) -> list[int]:
```
this function is unused in this script . The _get_subset_info function seems to be sufficient.

sjfleming replied:
excellent thanks!

magrenimish commented on tests/test_socam.py:
tests the compute_valid_mask function from the socam script which is unused.

sjfleming replied:
will get rid of that test

magrenimish commented on cellarium/ml/distributions/__init__.py:
```diff
@@ -2,5 +2,6 @@
+from cellarium.ml.distributions.unnormalized_categorical import PyroCategorical
 
-__all__ = ["NegativeBinomial"]
+__all__ = ["NegativeBinomial", "PyroCategorical"]
```
should this be renamed to "UnnormaizedPyroCategorical" to avoid confusion with regular PyroCategorical?

sjfleming replied:
Yes, I'll call it `UnconstrainedPyroCategorical` I think

sjfleming committed:
draft config file

sjfleming committed:
Merge branch 'main' into socam_integration

sjfleming committed:
code review fixes

sjfleming committed:
allow convert_to_tensor to handle pandas series

sjfleming committed:
update config

sjfleming committed:
cast possible bool input as float

sjfleming committed:
replace underscore names with colon names (matches cellxgene)

sjfleming committed:
remove scheduler for now

sjfleming committed:
add "unknown" to cell type categories in helpers

sjfleming committed:
bug fix - unknown self connection

sjfleming committed:
float dtypes

sjfleming committed:
make divide by scale float

sjfleming committed:
add float protection to normalize_total

sjfleming committed:
divide_by_scale coerce nans to zero with warning

sjfleming committed:
fix device issue

sjfleming committed:
use argument linking for default cell type names subset

sjfleming committed:
fix test for socam

sjfleming committed:
factor out propagation code, test it, make memory efficient

sjfleming committed:
lint

sjfleming committed:
make private

sjfleming committed:
handle propagation math better, avoid nan values

sjfleming committed:
revert numerical problems while keeping memory low

sjfleming committed:
add compilation back in

sjfleming committed:
revert to original implementation of logsumexp_propagated

sjfleming commented:
Actually I think the way I'm handling subsetting of the ontology does not make sense.  I've been taking all cells annotated in the dataset as my subset.  But what if some ancestor nodes from the ontology are missing?  This will mess up probability propagation I think!

sjfleming committed:
bug fix: include all ancestor nodes of user requested cell type nodes

sjfleming committed:
potential speedup: leaf versus nonleaf

sjfleming committed:
attempted fix for compilation

sjfleming committed:
try compiling regression math

sjfleming committed:
working example config file

sjfleming committed:
remove pyro; subset to active cell types from start not on fly (#384)

sjfleming committed:
Merge branch 'main' into socam_integration

sjfleming committed:
put series conversion back in

sjfleming committed:
make socam test use sparse mode

sjfleming committed:
lint

sjfleming committed:
easy way out: only test socam CLI single device

sjfleming committed:
change NaN replacement in divide_by_scale to suppress NaN genes

sjfleming committed:
remove kwargs from forward

sjfleming commented:
Sparse dataloader strategy #387 takes SOCAM train time (on 1001 gzip-compressed persistent-disk local h5ad files) per epoch on a T4 gpu from 20 mins to 5 mins.

fedorgrab commented:
> Sparse dataloader strategy #387 takes SOCAM train time (on 1001 gzip-compressed persistent-disk local h5ad files) per epoch on a T4 gpu from 20 mins to 5 mins.

That's incredible!

sjfleming committed:
log weights histogram once per epoch rather than every batch

sjfleming committed:
fix kwargs problem with validate method

sjfleming committed:
update config

sjfleming committed:
update from #390 for testing

sjfleming committed:
update val check frequency

sjfleming committed:
update to #390 again

sjfleming committed:
update val_check_interval default

sjfleming committed:
add lambda to config

sjfleming reviewed (approved):
Looks good now!
